pymc3/distributions/distribution.py

```diff
@@ -141,26 +159,102 @@ def __new__(cls, name, *args, **kwargs):
 
+        # Pop out PyMC3-related kwargs so only the disttribution kwargs remain
+        rng = kwargs.pop("rng", None)
         if rng is None:
             rng = model.default_rng
 
-        if not isinstance(name, string_types):
-            raise TypeError(f"Name needs to be a string but got: {name}")
-
         data = kwargs.pop("observed", None)
-
         total_size = kwargs.pop("total_size", None)
+        testval = kwargs.pop("testval", None)
+        transform = kwargs.pop("transform", UNSET)
 
+        shape = kwargs.pop("shape", None)
         dims = kwargs.pop("dims", None)
+        size = kwargs.pop("size", None)
+
+        # Raise on unsupported parametrization
+        if shape is not None and dims is not None:
+            raise ValueError("Passing both `shape` ({shape}) and `dims` ({dims}) is not supported!")
+        if dims is not None and size is not None:
+            raise ValueError("Passing both `dims` ({dims}) and `size` ({size}) is not supported!")
+        if shape is not None and size is not None:
+            raise ValueError("Passing both `shape` ({shape}) and `size` ({size}) is not supported!")
+
+        # Warn about discouraged parametrization
+        if shape is not None and not isinstance(shape, (list, tuple)):
+            warnings.warn("The `shape` parameter should be a list or tuple.", UserWarning)
+            shape = (shape,)
+        if dims is not None and not isinstance(dims, (list, tuple)):
+            warnings.warn("The `dims` parameter should be a list or tuple.", UserWarning)
+            dims = (dims,)
+        if size is not None and not isinstance(size, (list, tuple)):
+            warnings.warn("The `size` parameter should be a list or tuple.", UserWarning)
+            size = (size,)
+
+        if size is not None and Ellipsis in size:
+            raise ValueError("The `size` parameter cannot contain an Ellipsis. Actual: {size}")
+        if not _valid_ellipsis_position(shape):
+            raise ValueError(
+                f"Ellipsis in `shape` may only appear in the last position. Actual: {shape}"
+            )
+        if not _valid_ellipsis_position(dims):
+            raise ValueError(
+                f"Ellipsis in `dims` may only appear in the last position. Actual: {dims}"
+            )
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        # Create the RV without specifying size or testval.
+        # The size will be expanded later (if necessary) and only then the testval fits.
+        rv_native = cls.dist(*args, rng=rng, testval=None, size=None, **kwargs)
+        implied_ndim = rv_native.ndim
+        implied_dims = ...  # TODO: infer dimension names from Variables!
+
+        if shape is None and dims is None and size is None:
+            size = ()
+        elif size is not None:
+            # User already specified how to expand the RV
+            pass
```

Comment:
There shouldn't be a condition for a no-op.

Reply:
> This introduces a considerable amount of complexity into `Distribution.__new__`; the `shape` logic really needs its own function/method, at the very least.

Absolutely. This is still a draft and the `.dist()` API is also on the ToDo list. For that reason too, the logic will move into a helper function.

> Also, this PR unnecessarily reinstates the use of `shape` internally (e.g. numerous change from `size` to `shape` in the tests), which is the **opposite** of what we set out to do.

Many tests trigger `UserWarning`s or use the API in another discouraged fashion. Where it makes sense, they should be refactored to use the API in a best-practice way.
In many tests that will still be `shape`, because they also pass a `testval`, making the test shape-explicit.
I didn't want to invest that time until we have pinned down what exactly are the unsupported/discouraged API flavors.